### ISSUE TITLE
[dynamo] Data-dep control-flow — residual ~773 breaks (LayerDrop @ #77, mask==1 @ #78 already split out)

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#122]
## ⚠️ SCOPE UPDATE 2026-05-11

This umbrella's effective scope has been REDUCED. When this issue was filed (2026-04-29) the dedup search missed two existing issues that already track the largest patterns inside it:

- **`#77`** (filed 2026-04-29) tracks the **HF transformers LayerDrop idiom** (`if self.training: if torch.rand([]) < self.layerdrop: continue`). Audit of this umbrella's named source sites against the LayerDrop pattern: 6 of the top-9 sites (`opt:375`, `bart:660`, `bart:537`, `blenderbot:623`, `blenderbot_small:591`, `bigbird_pegasus:1834`) are LayerDrop occurrences, totaling **56 breaks across 6 sites**. Action for these: tracked separately at #77.
- **`#78`** (filed 2026-04-29) tracks the **`torch.all(attention_mask == 1)` linear-attn / mamba mask fast-path**. 3 of the top-9 sites (`qwen3_5:1312`, `granitemoehybrid:1219`, `qwen3_5_moe:1437`) are this pattern, totaling **28 breaks across 3 sites**. Action for these: tracked separately at #78.

**Current effective scope of THIS umbrella (post-audit):** ~773 breaks not covered by #77 or #78. Of those:
- **~520 breaks at "unknown" source locations** — blocked on a corpus-side improvement to `sweep/worker.py` to capture richer break-reason context. Once that lands, these can be classified and tracked at named-pattern issues (filed separately or merged into existing issues as appropriate).
- **~253 breaks at long-tail named source sites** (each ≤5 breaks per site, tail not yet enumerated).

The model table below remains as historical evidence but is **partially superseded by #77 + #78 for the LayerDrop train-mode entries**. Please consult those issues first before treating any train-mode-only row here as new signal.

**For Dynamo team:** action lives at #77 and #78 for the patterns called out above. This umbrella's marginal value is the long-tail enumeration, blocked on the corpus-side worker.py upgrade.

---

## Summary

**857 graph breaks across 129 distinct models** caused by Dynamo not supporting tracing of dynamic control flow (data-dependent `if`/`while` branching). Per the SCOPE UPDATE above, the LayerDrop and mask-all-equal-1 sub-patterns are now tracked at #77 and #78 respectively; this umbrella covers the residual ~773 breaks.

Numbers below are from the corpus 2026-04-29 sweep of HF transformers + diffusers + custom suites; per-model break counts come from the sweep's `explain_results.json`. The named-source attribution in the "Top source locations" block was added 2026-05-11 by reading the cited transformers source lines and matching them against the LayerDrop / mask==1 patterns.

## Break Reason

```
Detected data-dependent branching (e.g. `if my_tensor.sum() > 0:`).
Dynamo does not support tracing dynamic control flow.
Hint: Use torch.cond to express dynamic control flow.
```

**Total at filing time: 857 breaks across 129 distinct models, 163 (model,mode) pairs**

| Model | eval | train |
|---|---|---|
| AudioFlamingo3ForConditionalGeneration | - | 7 |
| AutoencoderKLHunyuanImageRefiner | 20 | 20 |
| AutoformerModel | - | 5 |
| BambaForCausalLM | 6 | 6 |
| BambaModel | 4 | 4 |
| BartForCausalLM | - | 5 |
| BartForConditionalGeneration | - | 9 |
| BartModel | - | 7 |
| BigBirdPegasusForCausalLM | - | 5 |
| BigBirdPegasusForConditionalGeneration | - | 3 |
| BigBirdPegasusModel | - | 3 |
| BioGptForCausalLM | - | 5 |
| BioGptModel | - | 3 |
| BlenderbotForCausalLM | - | 5 |
| BlenderbotForConditionalGeneration | - | 9 |
| BlenderbotModel | - | 7 |
| BlenderbotSmallForCausalLM | - | 5 |
| BlenderbotSmallForConditionalGeneration | - | 9 |
| BlenderbotSmallModel | - | 7 |
| Blip2ForConditionalGeneration | - | 7 |
| Blip2Model | - | 7 |
| CohereAsrForConditionalGeneration | - | 8 |
| CohereAsrModel | - | 6 |
| ConsistencyDecoderVAE | 1 | 1 |
| DFineModel | - | 7 |
| Data2VecAudioModel | - | 1 |
| EncodecModel | 21 | 21 |
| FSMTForConditionalGeneration | - | 5 |
| FSMTModel | - | 4 |
| FalconH1ForCausalLM | 6 | 6 |
| FalconH1Model | 4 | 4 |
| FastSpeech2ConformerModel | 2 | 2 |
| FlaubertModel | - | 1 |
| FlaubertWithLMHeadModel | - | 1 |
| Florence2ForConditionalGeneration | - | 6 |
| Florence2Model | - | 6 |
| GraniteMoeHybridForCausalLM | 6 | 6 |
| GraniteMoeHybridModel | 4 | 4 |
| GroundingDinoModel | - | 1 |
| HiDreamImageTransformer2DModel | 1 | 1 |
| HubertModel | - | 1 |
| IBertModel | - | 28 |
| InformerModel | - | 6 |
| InstructBlipForConditionalGeneration | - | 7 |
| InstructBlipModel | - | 7 |
| InstructBlipVideoForConditionalGeneration | - | 7 |
| InstructBlipVideoModel | - | 7 |
| JambaForCausalLM | 6 | 6 |
| JambaModel | 4 | 4 |
| Kosmos2ForConditionalGeneration | - | 7 |
| Kosmos2Model | - | 7 |
| LEDForConditionalGeneration | - | 3 |
| LEDModel | - | 3 |
| M2M100ForConditionalGeneration | - | 7 |
| M2M100Model | - | 6 |
| MBartForCausalLM | - | 5 |
| MBartForConditionalGeneration | - | 4 |
| MBartModel | - | 4 |
| MMGroundingDinoModel | - | 1 |
| MarianForCausalLM | - | 5 |
| MarianMTModel | - | 9 |
| MarianModel | - | 7 |
| MiniCPM-V-Resampler | 3 | 3 |
| MiniCPMV4_6ForConditionalGeneration | 8 | 8 |
| MiniCPMV4_6Model | 6 | 6 |
| MvpForCausalLM | - | 2 |
| MvpForConditionalGeneration | - | 5 |
| MvpModel | - | 4 |
| NemotronHForCausalLM | 6 | 6 |
| NemotronHModel | 4 | 4 |
| NllbMoeForConditionalGeneration | - | 8 |
| NllbMoeModel | - | 6 |
| OPTForCausalLM | - | 5 |
| OPTModel | - | 5 |
| OlmoHybridForCausalLM | 6 | 6 |
| OlmoHybridModel | 4 | 4 |
| OpenAIPrivacyFilterModel | 1 | 1 |
| OpenVoice-SynthesizerTrn-infer | 3 | - |
| Ovis2ForConditionalGeneration | 1 | 1 |
| PLBartForCausalLM | - | 5 |
| ... and 49 more | | |

Top source locations (audit attribution after 2026-05-11 split walk):
```
  520× unknown                                                                        — RESIDUAL (this umbrella; needs worker.py richer capture)
   16× transformers/models/opt/modeling_opt.py:375                                    — LayerDrop  → tracked at #77
   12× transformers/models/bart/modeling_bart.py:660                                  — LayerDrop  → tracked at #77
   12× transformers/models/qwen3_5/modeling_qwen3_5.py:1312                           — mask==1    → tracked at #78
   10× transformers/models/bart/modeling_bart.py:537                                  — LayerDrop  → tracked at #77
    8× transformers/models/granitemoehybrid/modeling_granitemoehybrid.py:1219         — mask==1    → tracked at #78
    8× transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py:1437                   — mask==1    → tracked at #78
    6× transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py:1834           — LayerDrop  → tracked at #77
    6× transformers/models/blenderbot/modeling_blenderbot.py:623                      — LayerDrop  → tracked at #77
    6× transformers/models/blenderbot_small/modeling_blenderbot_small.py:591          — LayerDrop  → tracked at #77
```

## Closure criterion (revised 2026-05-11)

This umbrella closes when the corpus-side `worker.py` improvement lands and the resulting ~520 currently-"unknown" breaks are classified into named patterns (filed separately or merged into existing issues as appropriate). The LayerDrop and mask==1 patterns no longer count toward this umbrella's scope — they're tracked at #77 and #78 respectively.

**Comment by penguinwu** [comment]
## Update — scope clarification (2026-05-04)

**This is the parent tracker for data-dependent control-flow branching across the corpus.** It's intentionally NOT split into 129 per-model issues — the umbrella-split policy says "same root cause at many sites stays as one issue, fix is the same per-site (`torch.cond` adoption or hoist branch outside compile region)."

### When per-pattern sub-issues are useful

When a SPECIFIC code-pattern of data-dep branching appears across many models with a SHARED upstream fix (single PR clears them all), file as a focused sub-issue. Existing examples:

- **#77** [dynamo] Data-dep branching: layerdrop pattern (89 models, 466 train-mode breaks)
- **#78** [dynamo] Data-dep branching: mamba/linear-attn mask all-ones fast-path (20 models, 204 breaks)

These two together account for ~78% of the 857 breaks counted here. They have shared upstream fix targets (HF refactor of layerdrop and the mamba mask construction).

### Remaining ~187 breaks not in #77 / #78

Distributed across many models with no single concentrated pattern. Top sites:
- `transformers/models/opt/modeling_opt.py:375` (16 breaks)
- `transformers/models/bart/modeling_bart.py:660` (12 breaks)
- `transformers/models/qwen3_5/modeling_qwen3_5.py:1312` (12 breaks)
- `transformers/models/bart/modeling_bart.py:537` (10 breaks)

If any of these crosses the threshold for a focused sub-issue (e.g., shared root cause across enough models to warrant a single PR fix), file under this umbrella as a sub-issue (similar to #77/#78).

### Closure

This umbrella closes when the number of data-dep branching breaks corpus-wide approaches zero — i.e., when `torch.cond` adoption or per-site refactors have eliminated them. Currently 857 corpus-wide. Track via:

```bash
python3 tools/sweep_compare.py --pattern "Detected data-dependent branching"
```
